File: apps/HealthApp/shared/FitFusionCore/Sources/FitFusionCore/Intelligence/BiologicalAgeEngine.swift
import Foundation

/// On-device heuristic biological-age estimator.
///
/// Inspired by published "biological age" approaches (PhenoAge, Klemera-Doubal,
/// resting-HR / HRV / VO\u{2082}Max regressions) but **simplified** so it runs in
/// pure Swift on a watch, with **no network, no third-party model, no medical
/// claim**. Each input nudges the chronological age up or down by a small,
/// transparent number of years; the breakdown is exposed to the UI so users
/// see *why* a number was assigned.
///
/// **Honest disclaimer:** this is an estimate for personal motivation, not
/// a clinical diagnosis. Don't make medical decisions from it.
public struct BiologicalAgeEngine {

    public static let shared = BiologicalAgeEngine()
    public init() {}

    // MARK: - Inputs

    public struct Inputs: Sendable, Hashable {
        public let chronologicalYears: Double
        public let sex: Sex
        public let restingHR: Double?           // bpm
        public let hrv: Double?                 // ms (SDNN)
        public let vo2Max: Double?              // ml/kg/min
        public let avgSleepHours: Double?       // h / night, 7-day avg
        public let bmi: Double?                 // kg/m\u{00b2}
        public let bodyFatPct: Double?          // 0\u{2013}1
        public let systolicBP: Double?          // mmHg
        public let diastolicBP: Double?         // mmHg
        public let weeklyExerciseMin: Double?   // min / week
        public let stepsPerDay: Double?
        public let smoker: Bool
        public let heavyAlcohol: Bool

        public init(chronologicalYears: Double, sex: Sex,
                    restingHR: Double? = nil, hrv: Double? = nil,
                    vo2Max: Double? = nil, avgSleepHours: Double? = nil,
                    bmi: Double? = nil, bodyFatPct: Double? = nil,
                    systolicBP: Double? = nil, diastolicBP: Double? = nil,
                    weeklyExerciseMin: Double? = nil, stepsPerDay: Double? = nil,
                    smoker: Bool = false, heavyAlcohol: Bool = false) {
            self.chronologicalYears = chronologicalYears
            self.sex = sex
            self.restingHR = restingHR
            self.hrv = hrv
            self.vo2Max = vo2Max
            self.avgSleepHours = avgSleepHours
            self.bmi = bmi
            self.bodyFatPct = bodyFatPct
            self.systolicBP = systolicBP
            self.diastolicBP = diastolicBP
            self.weeklyExerciseMin = weeklyExerciseMin
            self.stepsPerDay = stepsPerDay
            self.smoker = smoker
            self.heavyAlcohol = heavyAlcohol
        }
    }

    public enum Sex: String, Codable, Sendable {
        case male, female, other
    }

    // MARK: - Output

    public struct Result: Sendable, Hashable {
        public let chronologicalYears: Double
        public let biologicalYears: Double
        public let confidence: Double           // 0\u{2013}1 \u{2014} scales with #signals available
        public let factors: [Factor]

        public var deltaYears: Double { biologicalYears - chronologicalYears }
        public var verdict: String {
            switch deltaYears {
            case ..<(-3): return "Significantly younger than your age \u{1F680}"
            case (-3)..<(-0.5): return "Younger than your age \u{2728}"
            case (-0.5)..<0.5: return "Right on track"
            case 0.5..<3: return "Slightly older than your age"
            case 3..<7: return "Notably older \u{2014} worth attention"
            default: return "Much older \u{2014} consider lifestyle changes \u{26A0}\u{fe0f}"
            }
        }
    }

    public struct Factor: Sendable, Hashable, Identifiable {
        public let id = UUID()
        public let name: String
        public let value: String      // human-readable input value, e.g. "62 bpm"
        public let deltaYears: Double // negative = makes you younger
        public let direction: Direction

        public enum Direction: String, Sendable { case better, neutral, worse }
    }

    // MARK: - Compute

    public func estimate(_ inputs: Inputs) -> Result {
        var factors: [Factor] = []

        if let rhr = inputs.restingHR {
            // Each bpm above 60 adds ~0.15 yr; below 60 subtracts up to ~3 yr.
            let delta = (rhr - 60) * 0.15
            factors.append(.init(
                name: "Resting HR",
                value: "\(Int(rhr)) bpm",
                deltaYears: clamp(delta, -3, 5),
                direction: deltaSign(delta)
            ))
        }
        if let hrv = inputs.hrv {
            // Higher HRV = younger. Sweet spot ~50 ms; each 10 ms above subtracts ~0.6 yr.
            let delta = -((hrv - 35) / 10.0) * 0.6
            factors.append(.init(
                name: "HRV (SDNN)",
                value: "\(Int(hrv)) ms",
                deltaYears: clamp(delta, -3, 4),
                direction: deltaSign(delta)
            ))
        }
        if let vo2 = inputs.vo2Max {
            // VO\u{2082}Max population norm depends on sex+age; rough penalty if low.
            let target = inputs.sex == .female ? 32.0 : 38.0
            let delta = -((vo2 - target) / 5.0) * 0.7
            factors.append(.init(
                name: "VO\u{2082} Max",
                value: String(format: "%.1f ml/kg/min", vo2),
                deltaYears: clamp(delta, -4, 5),
                direction: deltaSign(delta)
            ))
        }
        if let sleep = inputs.avgSleepHours {
            // Optimal ~7.5 h. Each hour off the optimal in either direction adds ~0.5 yr.
            let off = abs(sleep - 7.5)
            let delta = off * 0.5
            factors.append(.init(
                name: "Avg sleep",
                value: String(format: "%.1f h", sleep),
                deltaYears: clamp(delta, 0, 4),
                direction: off < 0.5 ? .better : (off < 1.0 ? .neutral : .worse)
            ))
        }
        if let bmi = inputs.bmi {
            // Healthy BMI 18.5\u{2013}24.9 = 0; outside that band penalises ~0.4 yr per unit.
            let delta: Double
            if bmi < 18.5 { delta = (18.5 - bmi) * 0.4 }
            else if bmi > 25 { delta = (bmi - 25) * 0.4 }
            else { delta = 0 }
            factors.append(.init(
                name: "BMI",
                value: String(format: "%.1f", bmi),
                deltaYears: clamp(delta, 0, 5),
                direction: delta > 0.5 ? .worse : .neutral
            ))
        }
        if let bf = inputs.bodyFatPct {
            // Healthy body fat depends on sex; rough heuristic.
            let target = inputs.sex == .female ? 0.25 : 0.18
            let delta = max(0, bf - target) * 25
            factors.append(.init(
                name: "Body fat",
                value: String(format: "%.0f%%", bf * 100),
                deltaYears: clamp(delta, 0, 4),
                direction: delta > 0.3 ? .worse : .neutral
            ))
        }
        if let sys = inputs.systolicBP {
            // Above 130 / under 90 worsens score.
            let delta: Double
            if sys > 130 { delta = (sys - 130) / 10.0 * 0.7 }
            else if sys < 90 { delta = (90 - sys) / 10.0 * 0.5 }
            else { delta = 0 }
            factors.append(.init(
                name: "Blood pressure",
                value: "\(Int(sys))/\(Int(inputs.diastolicBP ?? 0)) mmHg",
                deltaYears: clamp(delta, 0, 4),
                direction: delta > 0.5 ? .worse : .neutral
            ))
        }
        if let exMin = inputs.weeklyExerciseMin {
            // 150 min / week target.
            let delta = -(min(exMin, 300) - 150) / 50.0 * 0.6
            factors.append(.init(
                name: "Weekly exercise",
                value: "\(Int(exMin)) min / wk",
                deltaYears: clamp(delta, -2.5, 2.5),
                direction: deltaSign(delta)
            ))
        }
        if let steps = inputs.stepsPerDay {
            let delta = -(min(steps, 12000) - 7500) / 1500.0 * 0.4
            factors.append(.init(
                name: "Daily steps",
                value: "\(Int(steps))",
                deltaYears: clamp(delta, -1.5, 2),
                direction: deltaSign(delta)
            ))
        }
        if inputs.smoker {
            factors.append(.init(name: "Smoking", value: "Yes",
                                 deltaYears: 6, direction: .worse))
        }
        if inputs.heavyAlcohol {
            factors.append(.init(name: "Heavy alcohol", value: "Yes",
                                 deltaYears: 3, direction: .worse))
        }

        let totalDelta = factors.reduce(0) { $0 + $1.deltaYears }
        let bio = max(15, min(120, inputs.chronologicalYears + totalDelta))

        // Confidence is the share of signals provided (chronological + sex always count).
        let provided = factors.count
        let possible = 11.0   // 9 numerics + 2 booleans
        let confidence = max(0.2, min(1.0, Double(provided) / possible))

        return Result(
            chronologicalYears: inputs.chronologicalYears,
            biologicalYears: bio,
            confidence: confidence,
            factors: factors.sorted { abs($0.deltaYears) > abs($1.deltaYears) }
        )
    }

    // MARK: - Helpers

    private func clamp(_ v: Double, _ lo: Double, _ hi: Double) -> Double {
        min(max(v, lo), hi)
    }
    private func deltaSign(_ d: Double) -> Factor.Direction {
        d < -0.25 ? .better : (d > 0.25 ? .worse : .neutral)
    }
}

File: apps/HealthApp/android/core/src/main/java/com/myhealth/core/intelligence/BiologicalAgeEngine.kt
package com.myhealth.core.intelligence

import kotlin.math.abs
import kotlin.math.max
import kotlin.math.min

/**
 * On-device biological-age heuristic — Kotlin port of the Swift
 * `BiologicalAgeEngine`. Same weights, same disclaimers: not a medical claim.
 */
object BiologicalAgeEngine {

    enum class Sex { female, male, other }
    enum class Direction { better, neutral, worse }

    data class Inputs(
        val chronologicalYears: Double,
        val sex: Sex,
        val restingHR: Double? = null,
        val hrv: Double? = null,
        val vo2Max: Double? = null,
        val avgSleepHours: Double? = null,
        val bmi: Double? = null,
        val bodyFatPct: Double? = null,
        val systolicBP: Double? = null,
        val diastolicBP: Double? = null,
        val weeklyExerciseMin: Double? = null,
        val stepsPerDay: Double? = null,
        val smoker: Boolean = false,
        val heavyAlcohol: Boolean = false,
    )

    data class Factor(
        val name: String,
        val value: String,
        val deltaYears: Double,
        val direction: Direction,
    )

    data class Result(
        val chronologicalYears: Double,
        val biologicalYears: Double,
        val confidence: Double,
        val factors: List<Factor>,
    ) {
        val deltaYears: Double get() = biologicalYears - chronologicalYears
        val verdict: String get() = when {
            deltaYears < -3        -> "Significantly younger than your age 🚀"
            deltaYears < -0.5      -> "Younger than your age ✨"
            deltaYears < 0.5       -> "Right on track"
            deltaYears < 3         -> "Slightly older than your age"
            deltaYears < 7         -> "Notably older — worth attention"
            else                   -> "Much older — consider lifestyle changes ⚠️"
        }
    }

    fun estimate(i: Inputs): Result {
        val factors = mutableListOf<Factor>()

        i.restingHR?.let { rhr ->
            val d = clamp((rhr - 60) * 0.15, -3.0, 5.0)
            factors += Factor("Resting HR", "${rhr.toInt()} bpm", d, dir(d))
        }
        i.hrv?.let { hrv ->
            val d = clamp(-((hrv - 35) / 10.0) * 0.6, -3.0, 4.0)
            factors += Factor("HRV (SDNN)", "${hrv.toInt()} ms", d, dir(d))
        }
        i.vo2Max?.let { vo2 ->
            val target = if (i.sex == Sex.female) 32.0 else 38.0
            val d = clamp(-((vo2 - target) / 5.0) * 0.7, -4.0, 5.0)
            factors += Factor("VO₂ Max", "%.1f ml/kg/min".format(vo2), d, dir(d))
        }
        i.avgSleepHours?.let { sleep ->
            val off = abs(sleep - 7.5)
            val d = clamp(off * 0.5, 0.0, 4.0)
            val direction = if (off < 0.5) Direction.better
                else if (off < 1.0) Direction.neutral else Direction.worse
            factors += Factor("Avg sleep", "%.1f h".format(sleep), d, direction)
        }
        i.bmi?.let { bmi ->
            val d = when {
                bmi < 18.5 -> (18.5 - bmi) * 0.4
                bmi > 25 -> (bmi - 25) * 0.4
                else -> 0.0
            }
            factors += Factor("BMI", "%.1f".format(bmi), clamp(d, 0.0, 5.0),
                if (d > 0.5) Direction.worse else Direction.neutral)
        }
        i.bodyFatPct?.let { bf ->
            val target = if (i.sex == Sex.female) 0.25 else 0.18
            val d = max(0.0, bf - target) * 25
            factors += Factor("Body fat", "${(bf * 100).toInt()}%", clamp(d, 0.0, 4.0),
                if (d > 0.3) Direction.worse else Direction.neutral)
        }
        i.systolicBP?.let { sys ->
            val d = when {
                sys > 130 -> (sys - 130) / 10.0 * 0.7
                sys < 90 -> (90 - sys) / 10.0 * 0.5
                else -> 0.0
            }
            factors += Factor("Blood pressure",
                "${sys.toInt()}/${(i.diastolicBP ?: 0.0).toInt()} mmHg",
                clamp(d, 0.0, 4.0),
                if (d > 0.5) Direction.worse else Direction.neutral)
        }
        i.weeklyExerciseMin?.let { exMin ->
            val d = clamp(-(min(exMin, 300.0) - 150) / 50.0 * 0.6, -2.5, 2.5)
            factors += Factor("Weekly exercise", "${exMin.toInt()} min / wk", d, dir(d))
        }
        i.stepsPerDay?.let { steps ->
            val d = clamp(-(min(steps, 12000.0) - 7500) / 1500.0 * 0.4, -1.5, 2.0)
            factors += Factor("Daily steps", steps.toInt().toString(), d, dir(d))
        }
        if (i.smoker) factors += Factor("Smoking", "Yes", 6.0, Direction.worse)
        if (i.heavyAlcohol) factors += Factor("Heavy alcohol", "Yes", 3.0, Direction.worse)

        val total = factors.sumOf { it.deltaYears }
        val bio = max(15.0, min(120.0, i.chronologicalYears + total))
        val possible = 11.0
        val confidence = max(0.2, min(1.0, factors.size / possible))

        return Result(
            chronologicalYears = i.chronologicalYears,
            biologicalYears = bio,
            confidence = confidence,
            factors = factors.sortedByDescending { abs(it.deltaYears) }
        )
    }

    private fun clamp(v: Double, lo: Double, hi: Double) = min(max(v, lo), hi)
    private fun dir(d: Double): Direction = when {
        d < -0.25 -> Direction.better
        d > 0.25 -> Direction.worse
        else -> Direction.neutral
    }
}
